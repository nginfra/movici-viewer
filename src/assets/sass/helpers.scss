.is-size-6-half {
  font-size: 0.85rem !important;
}

.text-ellipsis {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: pre;
}

// .x-overflow is used alongside with FixedPosition.ts as the overflow parent
//
.overflow,
.x-overflow {
  overflow: auto;
  padding-right: 0.5rem;
  margin-right: -0.5rem;
}

.overflow-hidden {
  overflow: hidden !important;
}

.overflow-hover {
  overflow: hidden;
  &:hover {
    overflow: auto;
    padding-right: 0.5rem;
    margin-right: -0.5rem;
  }
}

button {
  &.is-borderless {
    border: 0;
  }
  &.is-transparent {
    background-color: transparent;
    color: $black;
  }
  &.is-round {
    border-radius: 100%;
  }
  &.movici {
    background: $white;
    color: $primary;
    border: 2px solid $white-ter !important;
    &:hover {
      color: $primary;
      background: $primary-light;
      border-color: $grey-light;
    }
    &:active {
      color: $primary;
      border-color: $primary !important;
    }
  }
}

[required] {
  label,
  .label {
    &::after {
      margin-left: 2px;
      content: '*';
      color: $grey;
      font-weight: bold;
    }
  }
}
