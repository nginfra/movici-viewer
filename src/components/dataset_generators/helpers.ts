// @ts-nocheck
import GeoJSON from 'ol/format/GeoJSON';
import { transformGeoJsonToCRS } from '@/crs';
import { MoviciError } from '@/errors';
import { DatasetGeneratorDataset } from '@/types';

export function findDataset(type: string, currentDataset: DatasetGeneratorDataset) {
  let currentChild, result;

  if (type === currentDataset.type) {
    return currentDataset;
  } else {
    if (currentDataset.childs === undefined) {
      return null;
    }

    // Loop over child types
    for (let childTypeIdx = 0; childTypeIdx < currentDataset.childs.length; ++childTypeIdx) {
      const currentChildType = currentDataset.childs[childTypeIdx];

      for (let childIdx = 0; childIdx < currentChildType.datasets.length; ++childIdx) {
        currentChild = currentChildType.datasets[childIdx];
        // Search in the current child
        result = findDataset(type, currentChild);

        if (result !== null) {
          return result;
        }
      }

      // Return the result if the node has been found
    }

    // The node has not been found and we have no more options
    return null;
  }
}

export function validateGeoJSON(textData: string) {
  let feature;
  try {
    feature = new GeoJSON().readFeature(textData);
  } catch (e) {
    console.warn('Invalid GeoJSON supplied');
    failMessage('Invalid GeoJSON supplied');
    return false;
  }

  const featureType = feature?.getGeometry()?.getType();
  if (featureType?.toLowerCase() == 'polygon') {
    return true;
  }

  failMessage(`Feature is not a Polygon but a ${featureType}`);
  return false;
}

export function transformGeoJSON(geojson: GeoJSON) {
  let transformed;
  try {
    transformed = transformGeoJsonToCRS(geojson);
  } catch (e) {
    console.warn(e.message);
    if (e instanceof MoviciError) {
      failMessage(e.message);
    }
  }
  return transformed;
}
