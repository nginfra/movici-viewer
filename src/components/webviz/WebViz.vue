<template>
  <div>
    <ExportModal
      :active="showExportModal"
      :scenario-uuid="currentScenarioUUID"
      :timestamp="timestamp"
      :timeline-info="timelineInfo"
      @close="showExportModal = false"
    ></ExportModal>
    <VizSidebar>
      <template v-slot:below-header="{ reduce }">
        <WebvizSettings :reduce="reduce" :value="settings" @input="setSettings($event)" />
      </template>
      <template v-slot:content="{ reduce }">
        <div class="is-flex is-flex-direction-column">
          <b-field class="is-flex-grow-1 is-flex-shrink-2">
            <LayerPicker v-model="layerInfos" :reduce="reduce" :mode="mode" @share="shareURL" />
          </b-field>
          <div class="buttons lower-buttons" :class="reduce ? 'is-centered' : 'is-right'">
            <BButton
              v-if="mode === 'scenario'"
              class="is-primary"
              @click="shareURL"
              icon-left="share-alt"
            >
              <template v-if="!reduce">Share URL</template>
            </BButton>
            <BButton class="is-primary" icon-left="download" @click="showExportModal = true">
              <template v-if="!reduce">{{ $t('actions.export') }}</template>
            </BButton>
          </div>
        </div>
      </template>
    </VizSidebar>
    <MapVis
      :layer-infos="validLayers"
      :timeline-info="timelineInfo"
      :view-state.sync="viewState"
      :timestamp.sync="timestamp"
      buildings
    >
      <template #control-right="{ map, onViewstateChange, basemap, setBasemap }">
        <SearchBar
          is-right
          :map="map"
          :view-state="viewState"
          @update:view-state="onViewstateChange($event)"
        />
        <NavigationControl is-right :value="viewState" @input="onViewstateChange($event)" />
        <BaseMapControl is-right :value="basemap" @input="setBasemap" />
      </template>
      <template #control-bottom="{ updateTimestamp }">
        <TimeSlider
          v-if="timelineInfo"
          :value="timestamp"
          @input="updateTimestamp($event)"
          :timeline-info="timelineInfo"
        />
      </template>
    </MapVis>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue, Watch } from 'vue-property-decorator';
import Deck from '@/components/webviz/Deck.vue';
import Buildings from '@/components/webviz/mapLayers/Buildings.vue';
import VizSidebar from '@/components/webviz/VizSidebar.vue';
import LayerPicker from '@/components/webviz/LayerPicker.vue';
import WebvizSettings from '@/components/webviz/WebvizSettings.vue';
import SearchBar from '@/components/webviz/controls/SearchBar.vue';
import NavigationControl from '@/components/webviz/controls/NavigationControl.vue';
import BaseMapControl from '@/components/webviz/controls/BaseMapControl.vue';
import TimeSlider from '@/components/webviz/TimeSlider.vue';

import {
  CameraOptions,
  EntityGroupSummary,
  Nullable,
  Project,
  TimeOrientedSimulationInfo,
  UUID,
  ViewConfig,
  VisualizationMode,
  VisualizationSettings
} from '@/types';
import WebvizStore from '@/store/modules/webviz';
import {
  datasetValidator,
  entityGroupValidator,
  layerInfosToLayerConfig,
  layerKindValidator,
  simplifiedCamera
} from '@/visualizers/viewHelpers';
import { copyToClipboard, getBaseURL, getEntitySummary } from '@/utils';
import { successMessage } from '@/snackbar';
import MapVis from '@/components/webviz/MapVis.vue';
import defaults from '@/components/webviz/defaults';
import isError from 'lodash/isError';

import { VisualizerInfo } from '@/visualizers';
import ExportModal from '@/components/webviz/ExportModal.vue';
import ProjectStore from '@/store/modules/ProjectStore';

@Component({
  name: 'WebViz',
  components: {
    Buildings,
    Deck,
    ExportModal,
    LayerPicker,
    SearchBar,
    NavigationControl,
    BaseMapControl,
    TimeSlider,
    VizSidebar,
    WebvizSettings,
    MapVis
  }
})
export default class WebViz extends Vue {
  @Prop([Object]) readonly viewConfig!: Nullable<ViewConfig>;
  layerInfos: VisualizerInfo[] = [];
  settings: VisualizationSettings | null = null;
  currentProjectUUID: string | null = null;
  timestamp = 0;
  datasetInfoContentErrors: Record<string, string[]> = {};
  viewState: Nullable<CameraOptions> = defaults.viewState();
  showExportModal = false;

  get mode() {
    return WebvizStore.mode;
  }

  get timelineInfo(): TimeOrientedSimulationInfo | null {
    return WebvizStore.timelineInfo;
  }

  get project(): Project | null {
    return ProjectStore.activeProject;
  }

  get currentScenarioUUID(): string | null {
    if (this.settings?.mode !== VisualizationMode.SCENARIO) {
      return null;
    }
    return this.settings?.scenario?.uuid ?? null;
  }

  @Watch('layerInfos')
  determineDuplicateErrors(layerInfos: VisualizerInfo[]) {
    const ids: Set<string> = new Set();
    for (const info of layerInfos) {
      info.unsetError('duplicate');

      if (ids.has(info.id)) {
        info.setError('duplicate', 'Duplicate data layer detected');
      }
      ids.add(info.id);
    }
  }

  get validLayers() {
    return this.layerInfos.filter(i => !Object.values(i.errors).length);
  }

  mounted() {
    if (this.project) {
      this.settings = defaults.visualisationSettings(this.project);
    }
    if (this.viewConfig) {
      this.parseViewConfig(this.viewConfig);
    }
  }

  setSettings(settings: VisualizationSettings | null) {
    const oldVal = this.settings;
    this.settings = settings;
    if (oldVal) {
      if (
        oldVal.project?.uuid !== settings?.project?.uuid ||
        oldVal.scenario?.uuid !== settings?.scenario?.uuid
      ) {
        this.layerInfos = [];
      }
    }
    if (this.settings) {
      WebvizStore.useSettings(this.settings).then(() => {});
    }
  }

  @Watch('viewConfig')
  parseViewConfig(value: ViewConfig) {
    WebvizStore.parseViewConfig(value)
      .then(result => {
        this.determineContentErrors(result.layerInfos);
        this.determineDuplicateErrors(result.layerInfos);
        return result;
      })
      .then(result => {
        this.layerInfos = result.layerInfos;
        this.viewState = result.camera;
        this.settings = result.settings;
        this.timestamp = result.timestamp;
      });
  }

  async determineContentErrors(infos: VisualizerInfo[]) {
    for (const info of infos) {
      info.unsetError('content');
      try {
        datasetValidator()(info);
        const datasetSummary = await WebvizStore.getDatasetSummary(info.datasetUUID as UUID);
        if (!datasetSummary) {
          throw new Error('Error contacting server');
        }
        entityGroupValidator(datasetSummary)(info);
        const entitySummary = getEntitySummary(info.entityGroup, datasetSummary);
        layerKindValidator(entitySummary as EntityGroupSummary)(info);
      } catch (e) {
        if (isError(e)) {
          info.setError('content', e.message);
        }
      }
    }
  }

  shareURL() {
    if (!this.project || !this.settings?.scenario || !this.viewState) {
      return;
    }
    const payload: ViewConfig = {
      project_name: this.project.name,
      scenario_name: this.settings.scenario.name,
      data_layers: layerInfosToLayerConfig(this.layerInfos),
      camera: simplifiedCamera(this.viewState),
      timestamp: this.timestamp
    };
    copyToClipboard(getBaseURL() + '?view=' + btoa(JSON.stringify(payload)));
    successMessage(this.$t('messages.viewCopiedToClipboard') as string);
  }
}
</script>

<style scoped lang="scss">
.lower-buttons {
  padding: 10px 0;
  background: $white-ter;
}
::v-deep {
  .map-control-left {
    left: 310px !important;
  }
  .map-control-bottom {
    width: 100% !important;
  }
}
</style>
