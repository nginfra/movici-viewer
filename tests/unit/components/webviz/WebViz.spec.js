import WebViz from '@/components/webviz/WebViz';
import { createComponentWrapper, createStore } from '../../helpers';

describe('WebViz.vue', () => {
  let wrapper, store;

  beforeEach(() => {
    // best way so far to spy on module methods if they are called on mounted
    store = createStore({
      getters: {
        activeProject() {
          return {
            uuid: '<uuid>',
            name: 'some_project',
            display_name: 'Some Project'
          };
        }
      }
    });
    wrapper = createComponentWrapper(WebViz, { store }, { shallowMount: true });
  });
  it('renders correctly', async () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
