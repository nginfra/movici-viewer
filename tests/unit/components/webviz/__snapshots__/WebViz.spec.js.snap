// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebViz.vue renders correctly 1`] = `
<div>
  <exportmodal-stub
    timestamp="0"
  />
   
  <vizsidebar-stub />
   
  <mapvis-stub
    buildings="true"
    layerinfos=""
    timestamp="0"
    viewstate="[object Object]"
  />
</div>
`;
